service/config_provider.go

```diff
@@ -75,6 +75,29 @@ type configProvider struct {
 	watcher chan error
 }
 
+// ConfigProviderOption is an option to change the behavior of ConfigProvider
+// returned by NewConfigProvider()
+type ConfigProviderOption func(opts *configProvider)
+
+func WithConfigMapProviders(c map[string]configmapprovider.Provider) ConfigProviderOption {
+	return func(opts *configProvider) {
+		opts.configMapProviders = c
+	}
+}
+
+func WithCfgMapConverters(c []config.MapConverterFunc) ConfigProviderOption {
+	return func(opts *configProvider) {
+		opts.cfgMapConverters = c
+	}
+}
```

Comment:
```suggestion
func WithConfigMapConverters(c ...config.MapConverterFunc) ConfigProviderOption {
	return func(opts *configProvider) {
		opts.cfgMapConverters = append(opts.cfgMapConverters, c...)
	}
}
```

I think I would prefer to see these append to the config, allowing for multiple invocations of the options, possibly provided by separate packages.  The invocation of `NewConfigProvider()` ultimately controls the order in which they're applied.  A small change would be required there to set the default values after applying options if the list is empty, rather than setting it first and allowing it to be overwritten.

The same applies to `WithConfigMapProviders()`.  There order is not important, so multiple maps can be merged.

Reply:
I am confused about your comment: `I think I would prefer to see these append to the config, allowing for multiple invocations of the options, possibly provided by separate packages.` are you suggesting a global?

Even without the "default" problem, your suggestion does not trivially allow "no converters" case.

Also these append thing is unclear to me, I would rather provide only one "Converter" and a "MultiConverter" implementation if the list is a problem.